File: src/musikdroid/app/src/main/java/io/casey/musikcube/remote/ui/activity/WebSocketActivityBase.java
package io.casey.musikcube.remote.ui.activity;

import android.content.Context;
import android.content.SharedPreferences;
import android.media.AudioManager;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.view.KeyEvent;
import android.view.MenuItem;

import com.uacf.taskrunner.LifecycleDelegate;
import com.uacf.taskrunner.Runner;
import com.uacf.taskrunner.Task;

import io.casey.musikcube.remote.playback.PlaybackService;
import io.casey.musikcube.remote.playback.PlaybackServiceFactory;
import io.casey.musikcube.remote.websocket.WebSocketService;

public abstract class WebSocketActivityBase extends AppCompatActivity implements Runner.TaskCallbacks {
    private WebSocketService wss;
    private boolean paused = true;
    private LifecycleDelegate runnerDelegate;
    private PlaybackService playback;
    private SharedPreferences prefs;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setVolumeControlStream(AudioManager.STREAM_MUSIC);
        this.runnerDelegate = new LifecycleDelegate(this, this, getClass(), null);
        this.runnerDelegate.onCreate(savedInstanceState);
        this.wss = WebSocketService.getInstance(this);
        this.playback = PlaybackServiceFactory.instance(this);
        this.prefs = getSharedPreferences("prefs", Context.MODE_PRIVATE);
    }

    @Override
    protected void onPause() {
        super.onPause();
        this.runnerDelegate.onPause();
        this.wss.removeClient(getWebSocketServiceClient());
        this.playback.disconnect(getPlaybackServiceEventListener());
        this.paused = true;
    }

    @Override
    protected void onResume() {
        super.onResume();
        this.runnerDelegate.onResume();
        this.playback = PlaybackServiceFactory.instance(this);
        this.playback.connect(getPlaybackServiceEventListener());
        this.wss.addClient(getWebSocketServiceClient());
        this.paused = false;
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        this.runnerDelegate.onDestroy();
    }

    @Override
    protected void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        this.runnerDelegate.onSaveInstanceState(outState);
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        boolean streaming = prefs.getBoolean("streaming_playback", false);

        /* if we're not streaming we want the hardware buttons to go out to the system */
        if (!streaming) {
            if (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
                playback.volumeDown();
                return true;
            }
            else if (keyCode == KeyEvent.KEYCODE_VOLUME_UP) {
                playback.volumeUp();
                return true;
            }
        }

        return super.onKeyDown(keyCode, event);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onTaskCompleted(String s, long l, Task task, Object o) {

    }

    @Override
    public void onTaskError(String s, long l, Task task, Throwable throwable) {

    }

    protected final Runner runner() {
        return this.runnerDelegate.runner();
    }

    protected final boolean isPaused() {
        return this.paused;
    }

    protected final WebSocketService getWebSocketService() {
        return this.wss;
    }

    protected final PlaybackService getPlaybackService() {
        return this.playback;
    }

    protected abstract WebSocketService.Client getWebSocketServiceClient();
    protected abstract PlaybackService.EventListener getPlaybackServiceEventListener();
}

File: src/musikdroid/app/src/main/java/io/casey/musikcube/remote/ui/activity/WebSocketActivityBase.kt
package io.casey.musikcube.remote.ui.activity

import android.content.Context
import android.content.SharedPreferences
import android.media.AudioManager
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.KeyEvent
import android.view.MenuItem
import com.uacf.taskrunner.LifecycleDelegate
import com.uacf.taskrunner.Runner
import com.uacf.taskrunner.Task
import io.casey.musikcube.remote.Application
import io.casey.musikcube.remote.data.IDataProvider
import io.casey.musikcube.remote.injection.*
import io.casey.musikcube.remote.playback.PlaybackService
import io.casey.musikcube.remote.playback.PlaybackServiceFactory
import io.casey.musikcube.remote.ui.extension.hideKeyboard
import io.casey.musikcube.remote.websocket.Prefs
import io.casey.musikcube.remote.websocket.WebSocketService
import io.reactivex.disposables.CompositeDisposable
import javax.inject.Inject

abstract class WebSocketActivityBase : AppCompatActivity(), Runner.TaskCallbacks {
    protected var disposables = CompositeDisposable()
    private lateinit var runnerDelegate: LifecycleDelegate
    private lateinit var prefs: SharedPreferences
    private var paused = false
    @Inject lateinit var wss: WebSocketService
    @Inject lateinit var dataProvider: IDataProvider

    protected val component: ViewComponent =
        DaggerViewComponent.builder()
            .appComponent(Application.appComponent)
            .dataModule(DataModule())
            .build()

    override fun onCreate(savedInstanceState: Bundle?) {
        component.inject(this)

        super.onCreate(savedInstanceState)

        volumeControlStream = AudioManager.STREAM_MUSIC
        runnerDelegate = LifecycleDelegate(this, this, javaClass, null)
        runnerDelegate.onCreate(savedInstanceState)
        playbackService = PlaybackServiceFactory.instance(this)
        prefs = getSharedPreferences(Prefs.NAME, Context.MODE_PRIVATE)
    }

    override fun onPause() {
        hideKeyboard()

        super.onPause()

        dataProvider.detach()
        runnerDelegate.onPause()

        val playbackListener = playbackServiceEventListener
        if (playbackListener != null) {
            playbackService?.disconnect(playbackServiceEventListener!!)
        }

        val wssClient = webSocketServiceClient
        if (wssClient != null) {
            wss.removeClient(webSocketServiceClient!!)
        }

        paused = true
    }

    protected fun isPaused(): Boolean {
        return paused
    }

    override fun onResume() {
        super.onResume()

        dataProvider.attach()
        runnerDelegate.onResume()

        playbackService = PlaybackServiceFactory.instance(this)

        val playbackListener = playbackServiceEventListener
        if (playbackListener != null) {
            this.playbackService?.connect(playbackServiceEventListener!!)
        }

        val wssClient = webSocketServiceClient
        if (wssClient != null) {
            wss.addClient(webSocketServiceClient!!)
        }

        paused = false

        disposables.dispose()
        disposables = CompositeDisposable()
    }

    override fun onDestroy() {
        super.onDestroy()
        runnerDelegate.onDestroy()
        dataProvider.destroy()
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        runnerDelegate.onSaveInstanceState(outState)
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        val streaming = prefs.getBoolean(
            Prefs.Key.STREAMING_PLAYBACK, Prefs.Default.STREAMING_PLAYBACK)

        /* if we're not streaming we want the hardware buttons to go out to the system */
        if (!streaming) {
            if (keyCode == KeyEvent.KEYCODE_VOLUME_DOWN) {
                playbackService?.volumeDown()
                return true
            }
            else if (keyCode == KeyEvent.KEYCODE_VOLUME_UP) {
                playbackService?.volumeUp()
                return true
            }
        }

        return super.onKeyDown(keyCode, event)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onTaskCompleted(taskName: String, taskId: Long, task: Task<*, *>, result: Any) {

    }

    override fun onTaskError(s: String, l: Long, task: Task<*, *>, throwable: Throwable) {

    }

    protected fun getWebSocketService(): WebSocketService {
        return wss
    }

    protected var playbackService: PlaybackService? = null
        private set

    protected val runner: Runner
        get() = runnerDelegate.runner()

    protected fun reloadPlaybackService() {
        if (!isPaused() && playbackService != null) {
            val playbackListener = playbackServiceEventListener

            if (playbackListener != null) {
                playbackService?.disconnect(playbackServiceEventListener!!)
            }

            playbackService = PlaybackServiceFactory.instance(this)

            if (playbackListener != null) {
                playbackService?.connect(playbackServiceEventListener!!)
            }
        }
    }

    protected open val webSocketServiceClient: WebSocketService.Client?
        get() = null

    protected open val playbackServiceEventListener: (() -> Unit)?
        get() = null
}
